r/R/transforms.R

```diff
@@ -923,3 +923,57 @@ regress_out <- function(mat, latent_data, prediction_axis = c("row", "col")) {
     vars_to_regress = vars_to_regress
   )
 }
+
+#################
+# Normalizations
+#################
+
+#' Normalize a matrix using log normalization
+#' @param mat (IterableMatrix) Matrix to normalize
+#' @param scale_factor (numeric) Scale factor to multiply matrix by for log normalization
+#' @param add_one (logical) Add one to the matrix before log normalization
+#' @returns log normalized matrix.
+#' @export
+normalize_log <- function(mat, scale_factor = 1e4, add_one = TRUE) {
+  assert_is(mat, "IterableMatrix")
+  assert_is_numeric(scale_factor)
+  assert_true(is.logical(add_one))
+  assert_greater_than_zero(scale_factor)
+  mat <- mat * scale_factor
+  if (!add_one) mat <- mat - 1 
+  return(log1p(mat))
+}
+
+
+#' Normalize a `(features x cells)`` matrix using term frequency-inverse document frequency
+#' @param mat (IterableMatrix) to normalize
+#' @param feature_means (numeric) Means of the features to normalize by.  If no names are provided, then 
+#' each numeric value is assumed to correspond to the feature mean for the corresponding row of the matrix.
+#' Else, map each feature name to its mean value.
+#' @returns tf-idf normalized matrix.
+#' @export
+normalize_tfidf <- function(mat, feature_means = NULL, threads = 1L) {
```

Comment:
- Nice touch thinking of the name-matching option for `feature_means`
- I think we might want to follow the ArchR/Signac TF-IDF default formulas which include some logarithms. Perhaps you were thinking users would combine `normalize_log` with `normalize_tfidf` but I think it's better to keep each standalone
- We should also give the specific normalization formula here